src/aws-codeartifact/domain.ts

```diff
@@ -0,0 +1,335 @@
+ */
+export class Domain extends DomainBase implements IDomain {
+  /**
+   * Creates a Domain object from existing domain attributes.
+   *
+   * @param scope The parent construct.
+   * @param id The construct id.
+   * @param attrs The attributes of the domain to import.
+   */
+  public static fromDomainAttributes(scope: Construct, id: string, attrs: DomainAttributes): IDomain {
+    class Import extends DomainBase {
+      public readonly domainArn = attrs.domainArn;
+      public readonly domainName = attrs.domainName;
+      public readonly encryptionKey = attrs.encryptionKey;
+      public readonly domainOwner = attrs.domainOwner;
+      public readonly domainEncryptionKey = attrs.encryptionKey?.keyArn;
+      protected readonly policy?: PolicyDocument | undefined = undefined;
+    }
+
+    return new Import(scope, id);
+  }
+
+  /**
+   * Creates an IDomain object from an existing CodeArtifact domain ARN.
+   *
+   * @param scope The parent construct.
+   * @param id The construct id.
+   * @param domainArn - The ARN (Amazon Resource Name) of the existing CodeArtifact domain.
+   */
+  public static fromDomainArn(scope: Construct, id: string, domainArn: string): IDomain {
+    const domainResourceArnParts = Stack.of(scope).splitArn(domainArn, ArnFormat.SLASH_RESOURCE_NAME);
+    if (
+      domainResourceArnParts.resource !== 'domain' ||
+      domainResourceArnParts.account === undefined ||
+      domainResourceArnParts.resourceName === undefined
+    ) {
+      throw new Error(`Expected a domain ARN, but got ${domainArn}`);
+    }
+    return Domain.fromDomainAttributes(scope, id, {
+      domainArn,
+      domainName: domainResourceArnParts.resourceName,
+      domainOwner: domainResourceArnParts.account,
+    });
+  }
+
+  /**
+   * (internal) The CloudFormation resource representing this CodeArtifact domain.
+   */
+  protected cfnResource: CfnDomain;
+  /**
+   * The properties used to create the CloudFormation resource for this domain.
+   */
+  private cfnResourceProps: CfnDomainProps;
+
+  /**
+   * The ARN (Amazon Resource Name) of this CodeArtifact domain.
+   */
+  readonly domainArn: string;
+  /**
+   * The name of this CodeArtifact domain.
+   */
+  readonly domainName: string;
+  /**
+   * The AWS KMS encryption key associated with this domain, if any.
+   */
+  readonly encryptionKey?: IKey;
+  /**
+   * The AWS account ID that owns this domain.
+   */
+  readonly domainOwner: string;
+
+  /**
+   * The ARN of the key used to encrypt the Domain
+   */
+  readonly domainEncryptionKey?: string;
+
+  protected policy?: PolicyDocument;
+
+  constructor(scope: Construct, id: string, props: DomainProps) {
+    super(scope, id);
+    this.policy = new PolicyDocument();
+
+    const encryptionKey =
+      props.encryptionKey ??
+      new Key(this, 'Key', {
+        description: `Key for CodeArtifact Domain ${props.domainName}`,
+        alias: `codeartifact-domain/${props.domainName}`,
+      });
```

Comment:
This would always create a KMS key, but the properties in CFn are optional. In other words, wouldn't an AWS-managed key be used? Or is there a reason why it has to be a customer key?

Reply:
If a customer key is required, is it not necessary to have permission to access the key?

Reply:
You are correct. The key is optional. I read up on it and you are right a AWS managed key is used by default. I guess I added this in our internal version, because of compliance related best practices. 

Should I better remove this then?

Reply:
> If a customer key is required, is it not necessary to have permission to access the key or key policy?

I am not sure, if the Codeartifact service principal needs access. At least I haven't run into issues with our setup, although that might be overly permissive. You are correct, however, that the key policy could/should be kept in mind here (https://docs.aws.amazon.com/codeartifact/latest/ug/domain-create.html#create-domain-kms-key-policy-example)

Reply:
> Should I better remove this then?

I would prefer it to be removed, to leave a way for users not to specify a key. Users who want to be strict can specify a key.

Reply:
> I am not sure, if the Codeartifact service principal needs access. At least I haven't run into issues with our setup, although that might be overly permissive. You are correct, however, that the key policy could/should be kept in mind here (https://docs.aws.amazon.com/codeartifact/latest/ug/domain-create.html#create-domain-kms-key-policy-example)

I wonder if we should do about the following...

- Whether the key policy for the access from `codeartifact` should be set to the key passed by the user in our `Domain` Construct.
  - (direction for access: CodeArtifact -> KMS Key)
- Whether the `grant` method in the `Domain` construct needs to grant permissions for the key to grantee.
  - (direction for access: Other service -> CodeArtifact with KMS Key)
  - Now only access to CodeArtifact is granted.

@hoegertn @moltar 
Do you know anything about this?

Reply:
In my setup the Principal does not need to have access to the key in its policy. But I have a grant for the Org on the key policy itself. Should not be needed within the account.

Reply:
Okay, I removed the automatic creation of the key.